### DISCUSSION TITLE
Quick add multiple nodes (like a macro to add a set of nodes)

### DISCUSSION THREAD

**Discussion by Steve-Mcl** [node-red/designs#72] [Design Proposals]
Carrying on from here: https://discourse.nodered.org/t/fr-quick-add-multiple-nodes-ready-linked/52439/7 

A means of adding common/related sets of nodes

## demo
![chrome_kNnbFCXNYw](https://user-images.githubusercontent.com/44235289/163830579-c2f1e9ff-169a-4db3-9925-b997d3c32a14.gif)

## Proposal (open to re-definition)
* The name of this feature should be known as a "macro"
* Macros are to be defined in a "JSON source" (file / file storage / settings file)
* The JSON structure should permit defining of...
  * label
  * icon
  * color
  * nodes
    * An array of node types
  * wires
    * An array of wires to add after nodes are created
  * links
    * An array of links to add after nodes are created (link in/link out nodes)
  * properties
    * An array of properties to apply to nodes after they are created 
    
### Stretch features
* A popup to request values for all of the nodes/flows properties that were setup in the overrides ...
  * This would be selective / based on the value of the nodes "override" settings - to indicate user input is required.  e.g. A property override of `{node: x, changeable: { type: "string", "label": "Enter a name for the debug node", default: "Debug #"} }`

**Comment by Steve-Mcl** [comment]
## JSON format (proposal)

### JSON schema 
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Node-RED Macro",
  "description": "A flow macro for Node-RED",
  "type": "array",
  "items": {
    "type": "object",
    "required": ["nodes"],
    "properties": {
      "label": { "type": "string" },
      "color": { "type": ["string", "integer"] },
      "nodes": {
        "type": "array", "required": [], "minItems": 1,
        "items": { "type": "string", "minLength": 1 }
      },
      "links": {
        "type": "array",
        "items": {
          "type": "object",
          "required": [ "source", "target"],
          "properties": {
            "source": { "type": "integer", "minimum": 0 },
            "target": { "type": "integer", "minimum": 0 }
          }
        }
      },
      "wires": {
        "type": "array",
        "items": {
          "type": "object",
          "required": ["source","sourcePort","target"],
          "properties": {
            "source": { "type": "integer", "minimum": 0 },
            "sourcePort": { "type": "integer", "minimum": 0 },
            "target": { "type": "integer", "minimum": 0 }
          }
        }
      },
      "properties": {
        "type": "array",
        "items": {
          "type": "object",
          "required": ["node","property","value"],
          "properties": {
            "node": { "type": "integer", "minimum": 0 },
            "property": { "type": "string" },
            "value": { }
          }
        }
      }
    }
  }
}
```


## Samples

### Sample 1
```json
[
    {
        "label": "inject → function → debug",
        "nodes": [ "inject", "function", "debug" ],
        "wires": [
            { "source": 0, "sourcePort": 0, "target": 1 },
            { "source": 1, "sourcePort": 0, "target": 2 }
        ],
        "properties": [
            { "node": 0, "property": "name", "value": "my inject" },
            { "node": 1, "property": "name", "value": "edit me" },
            { "node": 2, "property": "name", "value": "result" }
        ]
    }
]
```

### Result 1
![image](https://user-images.githubusercontent.com/44235289/163818101-45d018a8-a263-4576-b5b8-7540403d7b9a.png)
![image](https://user-images.githubusercontent.com/44235289/163817956-bc835369-3001-4329-b044-417a96170aa7.png)


### Sample 2
```json
[
    {
        "label": "link call → link in → link return",
        "nodes": [ "link call", "link in", "link out" ],
        "links": [
            { "source": 0, "target": 1 }
        ],
        "wires": [
            { "source": 1, "sourcePort": 0, "target": 2 }
        ],
        "properties": [
            {"node": 2, "property": "mode", "value": "return"}
        ]
    }
]
```

### Result 2
![image](https://user-images.githubusercontent.com/44235289/163818234-afe8012b-686c-4ae4-a3a2-2509c2b765e0.png)

![image](https://user-images.githubusercontent.com/44235289/163818303-c6fd92e9-5e32-423a-b928-825f7c06d065.png)


### Sample 3
```json
[
    {
        "label": "Create http endpoint",
        "icon": "font-awesome/fa-wifi",
        "color": "#32a852",
        "nodes": [ "http in", "function", "http response" ],
        "wires": [
            { "source": 0, "sourcePort": 0, "target": 1 },
            { "source": 1, "sourcePort": 0, "target": 2 }
        ],
        "properties": [
            {"node": 0, "property":"url", "value": "/endpoint"},
            {"node": 0, "property":"name", "value": "endpoint"},
            {"node": 1, "property":"name", "value": "edit me"},
            {"node": 2, "property":"name", "value": "set my headers"}
        ]
    }
]
```

### Result 3

![image](https://user-images.githubusercontent.com/44235289/163827086-f86d1d14-8aaf-4c8a-8ecf-9804ab7eb173.png)

![image](https://user-images.githubusercontent.com/44235289/163827452-784678be-30cf-445a-b996-942a8f9cb58a.png)


---

Validated here: https://www.jsonschemavalidator.net/

**Comment by dceejay** [comment]
Devils advocate here.... as these look like tiny flows... should not the flow part of the schema be the existing flow format ? ie an array - and then wrap that into an object that holds the other metadata you need ... That may make "composing" macros easier in the future ?

**Reply by Steve-Mcl** [reply]
I have literally gone back and forwards on that idea and this one.
There were a bunch of considerations I made that landed me here (I forget what they are right now - i think it was internal implementation details)

However, what you say has merit - I will defiantly give it a go.

Thanks for the comments Dave.

**Reply by dceejay** [reply]
I guess it depends how far we intend to take this... How are the macros to be designed ? If they use the flow format then of course they could be created/edited with the editor. If they are only intended to be hand crafted snippets of 1-3 nodes then maybe you can get away with something else... 

(Not wishing to pollute your idea... but doing it anyway... - the sorts of things you are talking about re-using snippets and properties thereof would all be good for a re-spin of the library also...)

**Reply by knolleary** [reply]
Everything up to the Stretch feature is what the library functionality already provides. I think we should look at how the library could be integrated with the quick-add dialog so that library entries can be indexed and including in the search results. That would probably require changes to the library API in order to support providing a single endpoint to retrieve a full library catalogue - the current API does incremental requests for each 'directory' level.

If we don't do that, then the question is where are these macros coming from? And why isn't it the library mechanism we already have?

I think the main differentiator between what we already have and this proposal is what you have listed as the stretch feature - having some way to prompt the user for input to customise the flow before it is added. That could be achieved by adding an additional meta-data property to each node in the flow that has something to customise - rather than wrapping the whole flow in an object.

That would make this backwards compatible - older versions of NR would import the flow as normal and ignore that meta-data property.

**Comment by dceejay** [comment]
"adding an additional meta-data property to each node".... could this be achieved by (ahem) letting groups have metadata... and using the group feature to define a macro ? :-)

**Reply by knolleary** [reply]
That would be one way of doing it... but wouldn't be the way I'd pick ;)
